### DISCUSSION TITLE
When to take ready state?

### DISCUSSION THREAD

**Discussion by tisonkun** [tikv/raft-rs#507] [Q&A]
`Ready` has several pairs of methods to either borrow or take the inner state:

* `message` / `take_message`
* `read_states` / `take_read_states`
* `entries` / `take_entries`
* `committed_entries` / `take_committed_entries`
* `persisted_messages` / `take_persisted_messages`

All of them are `pub` methods and I don't know if I should always take the state and maintain myself or just borrow them on read-only case - does raft core processing logic depend on whether or not the state is taken?

cc @BusyJay

**Comment by BusyJay** [comment]
They are just accessors that allow read-only or move operations. raft doesn't not depend on or check if they are processed, it just always assume application has processed the ready following the protocol described in docs.

**Reply by tisonkun** [reply]
Thanks for your reply! I noticed that I should pass back the `Ready` struct on `advance*` methods so it can be a question.

But those methods seems look at soft state and hard state only, which is copied and not taken.
